komga/src/flyway/kotlin/db/migration/sqlite/V20210117175346__trash_bin_part_2.kt

```diff
@@ -0,0 +1,62 @@
+package db.migration.sqlite
+
+import mu.KotlinLogging
+import net.jpountz.xxhash.XXHashFactory
+import org.flywaydb.core.api.migration.BaseJavaMigration
+import org.flywaydb.core.api.migration.Context
+import org.springframework.jdbc.core.JdbcTemplate
+import org.springframework.jdbc.core.ResultSetExtractor
+import org.springframework.jdbc.datasource.SingleConnectionDataSource
+import java.io.FileInputStream
+import java.net.URL
+import java.nio.file.Path
+import java.nio.file.Paths
+
+private val logger = KotlinLogging.logger {}
+
+class V202101091156000__trash_bin_part_2 : BaseJavaMigration() {
+  override fun migrate(context: Context) {
+    val jdbcTemplate = JdbcTemplate(SingleConnectionDataSource(context.connection, true))
+
+    val totalCount = jdbcTemplate.queryForObject("select count(*) from BOOK", Long::class.java)
+    var currentCount = 0
+
+    logger.info { "Calculating hash for $totalCount books" }
+    while (existsRecordWithoutHash(jdbcTemplate)) {
+      val values = jdbcTemplate.queryForList("select b.ID, b.URL from BOOK b where b.FILE_HASH == '' limit 500")
+      currentCount += values.size
+
+      val params = values.map {
+        val hash = getHash(Paths.get((URL(it["URL"] as String).toURI())))
+        arrayOf(hash, it["ID"])
+      }.toList()
+      jdbcTemplate.batchUpdate("UPDATE BOOK SET FILE_HASH = ? WHERE ID = ?", params)
+      logger.info { "Hash calculation progress: $currentCount/$totalCount" }
+    }
+  }
+
+  fun getHash(path: Path): String {
+    val seed = -0x68b84d74L
+    val file = path.toFile()
+    val bytesToSkip = java.lang.Long.highestOneBit(file.length() / 100)
+
+    with(FileInputStream(file)) {
+      val buf = ByteArray(4096)
+      val hash64 = XXHashFactory.fastestInstance().newStreamingHash64(seed)
+
+      while (true) {
+        val read: Int = this.read(buf)
+        if (read == -1) break
+
+        hash64.update(buf, 0, read)
+        this.skip(bytesToSkip)
```

Comment:
Why do we need to skip some bytes?

Reply:
To speed up hash calculation. When calculating hash we are, for the most part, i/o bound. Reading entire file can be slow on network mounted drives or with large collections. Reading only parts of a file while comparing file sizes should be enough.
Here, rather roughly approximated, it reads 100 x 4KiB chunks in total, skipping ~1/100 size of a file after every read. 
I've done some benchmarking on a regular hdd with 25gb library and file sizes of around 500mb:
It takes ~134s when reading entire file and with skipping it takes ~27.3s which is 5 times faster.